# Assignment 1 - Shell basics - Jupyter Notebook I

Work through as much of the Software Carpentry [lesson on the Unix Shell](http://swcarpentry.github.io/shell-novice/) as you can.  Run through the Setup section just below, then open a shell from the command line or with a terminal session through Jupyter to run through the exercises.

After you have completed the first few sections of the tutorial, return to this notebook.

Execute all of the cells, answer all of the questions, and wherever you see "**Edit this cell**", do it!


## 0. Setup - getting required files 

To get started, you'll need to have the required files in your directory.  Use `wget` to get them:

In [3]:
!wget https://swcarpentry.github.io/shell-novice/data/shell-lesson-data.zip

--2023-09-11 02:15:48--  https://swcarpentry.github.io/shell-novice/data/shell-lesson-data.zip
Resolving swcarpentry.github.io (swcarpentry.github.io)... 185.199.109.153, 185.199.110.153, 185.199.111.153, ...
Connecting to swcarpentry.github.io (swcarpentry.github.io)|185.199.109.153|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 460289 (450K) [application/zip]
Saving to: ‘shell-lesson-data.zip.1’

shell-lesson-data.z 100%[===================>] 449.50K  --.-KB/s    in 0.02s   

2023-09-11 02:15:48 (23.4 MB/s) - ‘shell-lesson-data.zip.1’ saved [460289/460289]



The `unzip -l` option shows you what will be unpacked before you actually unpack it.  It's always a good habit to check what you're going to find before you fill your disk with new files.

In [4]:
!unzip -l shell-lesson-data.zip

Archive:  shell-lesson-data.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  2023-03-12 12:40   shell-lesson-data/
        0  2023-03-12 12:40   shell-lesson-data/north-pacific-gyre/
        0  2023-03-12 12:42   shell-lesson-data/exercise-data/
        0  2023-03-12 12:40   shell-lesson-data/exercise-data/writing/
  1053440  2023-03-12 12:40   shell-lesson-data/exercise-data/writing/LittleWomen.txt
      218  2023-03-12 12:40   shell-lesson-data/exercise-data/writing/haiku.txt
        0  2023-03-12 12:40   shell-lesson-data/exercise-data/creatures/
     1838  2023-03-12 12:40   shell-lesson-data/exercise-data/creatures/basilisk.dat
     1833  2023-03-12 12:40   shell-lesson-data/exercise-data/creatures/unicorn.dat
     1830  2023-03-12 12:40   shell-lesson-data/exercise-data/creatures/minotaur.dat
        0  2023-03-12 12:40   shell-lesson-data/exercise-data/animal-counts/
      155  2023-03-12 12:40   shell-lesson-data/exercise-data/animal-counts/a

Looks good - now we remove the `-l` so we can `unzip` it for real this time.

In [14]:
!unzip shell-lesson-data.zip

Archive:  shell-lesson-data.zip
replace shell-lesson-data/exercise-data/writing/LittleWomen.txt? [y]es, [n]o, [A]ll, [N]one, [r]ename: ^C


*Note*: you only need to do this once per session while using Jupyter.  You can open a terminal now and work through the steps, and return to this notebook a little later, and the files will be available either way.  That's because you're working in the same local directory.

If you're reusing one AWS EC2 instance, your files should remain after you stop and restart. 

## 1. Markdown basics (10%)

In the following cells, practice marking up text using markdown:

# Make this a top-level header (add "`# `" to the left of the text).

## Make this a second-level header (add "`## `" to the left of the text).

Make the following three words three separate bullet points (add "`* `" to the left of each line):

  `* ` One
  `* ` Two
  `* ` Three

Make this cell a Markdown cell instead of code!

In [6]:
print("Make this cell a code cell instead of Markdown!")

Make this cell a code cell instead of Markdown!


## 2. Navigating Files and Directories (10%)

As you work through this section of the tutorial, complete the steps here as well, using the `!` shell escape command.  Execute each cell as you go.

These steps aren't exactly the same as what's in the tutorial, where the file layout is a little different and where they're not using a notebook like we are.  That's okay.  Just consider this practice.

In [7]:
!whoami

ubuntu


In [8]:
!pwd

/home/ubuntu/notebooks/Assignment 1


In [9]:
!ls -F

'assignment-1-Fall-2023(1) (3).ipynb'   quotations.txt
 filelist.txt			        quotes.txt
 mydirectory/			        shell-lesson-data/
 myfiles.txt			        shell-lesson-data.zip
 partone/			        shell-lesson-data.zip.1


In [10]:
!ls -F shell-lesson-data/

exercise-data/	north-pacific-gyre/


In [11]:
!ls -aF

 ./				        partone/
 ../				        quotations.txt
 .ipynb_checkpoints/		        quotes.txt
'assignment-1-Fall-2023(1) (3).ipynb'   shell-lesson-data/
 filelist.txt			        shell-lesson-data.zip
 mydirectory/			        shell-lesson-data.zip.1
 myfiles.txt


In [12]:
!ls -af .

 myfiles.txt		  'assignment-1-Fall-2023(1) (3).ipynb'
 quotes.txt		   filelist.txt
 partone		   quotations.txt
 ..			   mydirectory
 shell-lesson-data.zip	   shell-lesson-data
 shell-lesson-data.zip.1   .ipynb_checkpoints
 .


What is the difference between the two previous cells, and what does the single dot mean?

**EDIT THIS CELL** WITH YOUR ANSWER HERE

**What is the difference between the two previous cells**
The differences are dot "." and uppercase and lowercase of "F" and "f". Furthermore, "f" is not able to sort and list the files and directories, but "F" is able in a default order

**what does the single dot mean?**
It lists the current directory for assigned content.

In [13]:
!ls -F ..

'Assignment 1'/      index.html.2		   shell-lesson-data.zip.1
'Untitled Folder'/   practice-01-2023.ipynb	   shell-lesson-data.zip.2
 Untitled.ipynb      practice-02-fall-2023.ipynb   shell-lesson-data.zip.3
 index.html	     shell-lesson-data/		   shell-lesson-data.zip.4
 index.html.1	     shell-lesson-data.zip	   untitled.txt


What do the double dots mean?

**EDIT THIS CELL** WITH YOUR ANSWER HERE

**What do the double dots mean?** It means to access to the parent directory of the current location

In [12]:
!ls shell-lesson-data/north-pacific-gyre/

NENE01729A.txt	NENE01751B.txt	NENE01971Z.txt	NENE02040A.txt	NENE02043B.txt
NENE01729B.txt	NENE01812A.txt	NENE01978A.txt	NENE02040B.txt	goodiff.sh
NENE01736A.txt	NENE01843A.txt	NENE01978B.txt	NENE02040Z.txt	goostats.sh
NENE01751A.txt	NENE01843B.txt	NENE02018B.txt	NENE02043A.txt


## 3. Working with Files and Directories (10%)

The following cells come from the next section of the tutorial.

In [16]:
!ls -F

'assignment-1-Fall-2023(1) (3).ipynb'   quotations.txt
 filelist.txt			        quotes.txt
 mydirectory/			        shell-lesson-data/
 myfiles.txt			        shell-lesson-data.zip
 partone/			        shell-lesson-data.zip.1


In [17]:
!mkdir thesis

In [30]:
import os
assert "thesis" in os.listdir()

In [16]:
!ls -F

'assignment-1-Fall-2023(1) (3).ipynb'   shell-lesson-data.zip
 shell-lesson-data/		        thesis/


You can't use the nano editor here in Jupyter, so we'll use the `touch` command to create an empty file instead.

In [17]:
!touch thesis/draft.txt

In [18]:
assert "draft.txt" in os.listdir("thesis")

In [19]:
!ls -F thesis

draft.txt


Removing files and directories.

In [20]:
!rm thesis/draft.txt

In [21]:
assert "draft.txt" not in os.listdir("thesis")

In [22]:
!rm thesis

rm: cannot remove 'thesis': Is a directory


In [23]:
!rmdir thesis

In [24]:
assert "thesis" not in os.listdir()

In [25]:
!ls

'assignment-1-Fall-2023(1) (3).ipynb'   shell-lesson-data.zip
 shell-lesson-data


Renaming and copying files.

In [26]:
!touch draft.txt

In [27]:
assert "draft.txt" in os.listdir()

In [28]:
!mv draft.txt quotes.txt

In [31]:
assert "quotes.txt" in os.listdir()
assert "draft.txt" not in os.listdir()

In [30]:
!ls

'assignment-1-Fall-2023(1) (3).ipynb'   shell-lesson-data
 quotes.txt			        shell-lesson-data.zip


In [32]:
!cp quotes.txt quotations.txt

In [33]:
assert "quotes.txt" in os.listdir()
assert "quotations.txt" in os.listdir()

## 4. Working with output redirection (20%)

Create a new directory:

In [34]:
!mkdir part1

Rename `part1` to `partone` using `mv`.

In [35]:
!mv part1 partone
!ls

'assignment-1-Fall-2023(1) (3).ipynb'   quotations.txt	 shell-lesson-data
 partone			        quotes.txt	 shell-lesson-data.zip


Create a file named `filelist.txt` using the output from `ls` and the output redirector `>`.

In [36]:
!ls > filelist.txt

In [37]:
!cat filelist.txt

assignment-1-Fall-2023(1) (3).ipynb
filelist.txt
partone
quotations.txt
quotes.txt
shell-lesson-data
shell-lesson-data.zip


Append to `filelist.txt` using the output appending redirector `>>`.  Note the difference between the single `>` and double `>>`.

In [40]:
!ls >> filelist.txt
!cat filelist.txt

assignment-1-Fall-2023(1) (3).ipynb
filelist.txt
partone
quotations.txt
quotes.txt
shell-lesson-data
shell-lesson-data.zip
assignment-1-Fall-2023(1) (3).ipynb
filelist.txt
partone
quotations.txt
quotes.txt
shell-lesson-data
shell-lesson-data.zip


In [41]:
!ls > filelist.txt
!cat filelist.txt

assignment-1-Fall-2023(1) (3).ipynb
filelist.txt
partone
quotations.txt
quotes.txt
shell-lesson-data
shell-lesson-data.zip


What's the difference between `>` and `>>`?


**EDIT THIS CELL** WITH YOUR ANSWER HERE

**What's the difference between `>` and `>>`?** For `>`, if the text file exist while coding it, than the new file will overite it. For `>>`, this will add on to the previous text file if the file exist while coding it.

Now create a directory called "`mydirectory`":

In [33]:
# Edit this cell!
!mkdir mydirectory

In [34]:
assert 'mydirectory' in os.listdir('.')

Using `ls` and output redirection, create a file called `myfiles.txt` in the directory `mydirectory` that contains the list of files in the current directory.

In [25]:
# Edit this cell!
!ls > mydirectory/myfiles.txt

In [26]:
assert 'myfiles.txt' in os.listdir('mydirectory')

Clean up the directory you just created by removing its contents (the file you created) using `rm`.

In [29]:
# Edit this cell!
!rm mydirectory/myfiles.txt

In [35]:
assert 'myfiles.txt' not in os.listdir('mydirectory')

In [36]:
!rmdir mydirectory

Now remove the directory itself using `rmdir`.

In [37]:
# Edit this cell!

In [38]:
assert 'mydirectory' not in os.listdir('.')

## 5. Filters and pipes (25%)

Let's look at something a little more interesting.  Download the text of Charlotte Bronte's *Jane Eyre* from [Project Gutenberg](http://www.gutenberg.org/):

In [3]:
!wget https://s3.amazonaws.com/2018-dmfa/assignment-1/jane-eyre.txt

--2023-09-11 15:43:31--  https://s3.amazonaws.com/2018-dmfa/assignment-1/jane-eyre.txt
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.217.87.102, 52.217.120.240, 16.182.35.240, ...
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.217.87.102|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1070331 (1.0M) [text/plain]
Saving to: ‘jane-eyre.txt.1’

jane-eyre.txt.1     100%[===================>]   1.02M  --.-KB/s    in 0.01s   

2023-09-11 15:43:31 (69.4 MB/s) - ‘jane-eyre.txt.1’ saved [1070331/1070331]



`head` and `tail` are very useful.  They let you take a quick peek at the start and end of files.

In [4]:
!head jane-eyre.txt

The Project Gutenberg eBook, Jane Eyre, by Charlotte Bronte, Illustrated
by F. H. Townsend


This eBook is for the use of anyone anywhere at no cost and with
almost no restrictions whatsoever.  You may copy it, give it away or
re-use it under the terms of the Project Gutenberg License included
with this eBook or online at www.gutenberg.org




In [5]:
!tail jane-eyre.txt

keep eBooks in compliance with any particular paper edition.

Most people start at our Web site which has the main PG search facility:

     http://www.gutenberg.org

This Web site includes information about Project Gutenberg-tm,
including how to make donations to the Project Gutenberg Literary
Archive Foundation, how to help produce our new eBooks, and how to
subscribe to our email newsletter to hear about new eBooks.


`grep` is one of the most useful filters.  It lets you search for and match lines that contain specific expressions.  For example, to find mentions of "copyright":

In [6]:
!grep copyright jane-eyre.txt

_in this Volume are the copyright of_
one owns a United States copyright in these works, so the Foundation
permission and without paying copyright royalties.  Special rules,
(trademark/copyright) agreement.  If you do not agree to abide by all
or PGLAF), owns a compilation copyright in the collection of Project
1.D.  The copyright laws of the place where you are located also govern
the copyright status of any work in any country outside the United
posted with permission of the copyright holder), the work can be copied
with the permission of the copyright holder, your use and distribution
terms imposed by the copyright holder.  Additional terms will be linked
permission of the copyright holder found at the beginning of this work.
effort to identify, do copyright research on, transcribe and proofread
corrupt data, transcription errors, a copyright or other intellectual
unless a copyright notice is included.  Thus, we do not necessarily


Notice anything that those lines have in common? Each line contains the word "copyright"

Let's add a little more information by including the `-n` flag to add matching line numbers.

In [7]:
!grep -n copyright jane-eyre.txt

57:_in this Volume are the copyright of_
20719:one owns a United States copyright in these works, so the Foundation
20721:permission and without paying copyright royalties.  Special rules,
20756:(trademark/copyright) agreement.  If you do not agree to abide by all
20775:or PGLAF), owns a compilation copyright in the collection of Project
20790:1.D.  The copyright laws of the place where you are located also govern
20797:the copyright status of any work in any country outside the United
20816:posted with permission of the copyright holder), the work can be copied
20826:with the permission of the copyright holder, your use and distribution
20828:terms imposed by the copyright holder.  Additional terms will be linked
20830:permission of the copyright holder found at the beginning of this work.
20901:effort to identify, do copyright research on, transcribe and proofread
20906:corrupt data, transcription errors, a copyright or other intellectual
21052:unless a copyright notice is included. 

Now let's look for any mention of "book".  This will match a lot of text, so we'll just take the first 10 matching lines by *piping* the output from `grep` into `head`.

In [8]:
!grep -n book jane-eyre.txt | head

101:doubt the tendency of such books as "Jane Eyre:" in whose eyes whatever
220:contained a bookcase: I soon possessed myself of a volume, taking care
229:book, I studied the aspect of that winter afternoon.  Afar, it offered a
234:I returned to my book--Bewick's History of British Birds: the letterpress
357:"Show the book."
361:"You have no business to take our books; you are a dependent, mama says;
365:rummage my bookshelves: for they _are_ mine; all the house belongs to me,
370:lift and poise the book and stand in act to hurl it, I instinctively
799:tart away.  Bessie asked if I would have a book: the word _book_ acted as
801:the library.  This book I had again and again perused with delight.  I
grep: write error: Broken pipe


How many lines contain "book"?  We can count by piping into the word count tool `wc`.

In [9]:
!grep book jane-eyre.txt | wc

     84    1055    5889


That's 84 matching lines, containing 1055 words and 5889 characters.  If you just wanted the lines by themselves, use `wc -l`:

In [10]:
!grep book jane-eyre.txt | wc -l

84


What if we want to match both upper- and lower-case text?  Use `grep -i`:

In [11]:
!grep time
jane-eyre.txt | wc -l

402


In [12]:
!grep -i time jane-eyre.txt | wc -l

402


How many lines in *Jane Eyre* contain "other" (just lower-case)?  Start by using `grep` to extract lines that match the word "other" in `jane-eyre.txt` and redirecting it to a file called `other-lines.txt`.

In [13]:
!grep other jane-eyre.txt > other-lines.txt

In [14]:
%sc h_other = head -1 other-lines.txt
assert "other" in h_other

In [15]:
%sc t_other = tail -1 other-lines.txt
assert "other" in t_other

Now count up the lines in the file you created using wc.

In [16]:
!wc -l other-lines.txt

426 other-lines.txt


In [17]:
!grep other other-lines.txt | wc

    426    5027   28621


Your answer should be 426!

## 6. Counting words with `grep` (25%)

By piping commands together we can do a lot of powerful things right at the command line.  Let's create a count of the most commonly occurring words in *Jane Eyre*.  To do that, we could write a Python or R script that just counts words, but with the command line shell tools we only need to put a proper pipeline together and we can often accomplish tasks like this in one line.

First we need to split up the text lines into a word per line.  There are `grep` flags for that!

In [18]:
!cat jane-eyre.txt | tr -sc '[:alpha:]' '[\n*]' | head -10


The
Project
Gutenberg
eBook
Jane
Eyre
by
Charlotte
Bronte
tr: write error: Broken pipe
tr: write error
cat: write error: Broken pipe


Now we need to sort them and count the unique tokens.  `sort` solves the first problem.

In [19]:
!cat jane-eyre.txt | tr -sc '[:alpha:]' '[\n*]' | sort | head -10


A
A
A
A
A
A
A
A
A
sort: write failed: 'standard output': Broken pipe
sort: write error


And `uniq -c` solves the second problem.

In [20]:
!cat jane-eyre.txt | tr -sc '[:alpha:]' '[\n*]' | sort | uniq -c | head -25

      1 
    159 A
      1 ACTUAL
      2 AGREE
      1 AGREEMENT
      1 AK
      1 AN
      1 AND
      3 ANY
      1 ANYTHING
      1 AS
      3 ASCII
      1 AUTHOR
      1 AUTOBIOGRAPHY
     25 Abbot
      1 Abhorred
      1 Abigail
      1 Abominable
      5 About
      3 Above
      1 Absent
      1 Absolutely
      1 According
      1 Accordingly
      1 Accustomed
uniq: write error: Broken pipe


But there's a catch... do you see it?

We need to convert all the words down into lower case so that we are correctly counting unique words.  There's another command, `tr`, for that.

In [21]:
!cat jane-eyre.txt | tr -sc '[:alpha:]' '[\n*]' | tr '[:upper:]' '[:lower:]' | sort | uniq -c | head -25

      1 
   4541 a
      3 abandon
      8 abandoned
      2 abandonment
      1 abate
     25 abbot
      4 abhor
      5 abhorred
      3 abide
      1 abigail
      2 abigails
      1 abilities
     23 able
      8 abode
      1 abodes
      2 abominable
      1 aboon
    228 about
     41 above
      1 abridge
      4 abroad
      8 abrupt
      6 abruptly
      1 abruptness
uniq: write error: Broken pipe


...and if we want to know only the top 10 words in *Jane Eyre*, we need to sort the output.

In [1]:
!cat jane-eyre.txt | tr -sc '[:alpha:]' '[\n*]' | tr '[:upper:]' '[:lower:]' | sort | uniq -c | sort | head -10

      1 
      1 abate
      1 abigail
      1 abilities
      1 abodes
      1 aboon
      1 abridge
      1 abruptness
      1 absences
      1 absorbing
sort: write failed: 'standard output': Broken pipe
sort: write error


But that sorts by character, not number.  Fortunately, `sort -n` does what we want.

In [2]:
!cat jane-eyre.txt | tr -sc '[:alpha:]' '[\n*]' | tr '[:upper:]' '[:lower:]' | sort | uniq -c | sort -n | head -10

      1 
      1 abate
      1 abigail
      1 abilities
      1 abodes
      1 aboon
      1 abridge
      1 abruptness
      1 absences
      1 absorbing
sort: write failed: 'standard output': Broken pipe
sort: write error


But that's the wrong end of the list!  Two ways to fix that:  (a) use `tail` instead of `head`; (b) use `sort -rn`, which will sort in reverse order.  Let's try the latter.

In [24]:
!cat jane-eyre.txt | tr -sc '[:alpha:]' '[\n*]' | tr '[:upper:]' '[:lower:]' | sort | uniq -c | sort -rn | head -10

   8033 the
   7268 i
   6699 and
   5318 to
   4541 a
   4484 of
   3066 you
   2826 in
   2527 was
   2431 it
sort: write failed: 'standard output': Broken pipe
sort: write error


Let's try another text.

Download *Through the Looking Glass* from https://s3.amazonaws.com/2018-dmfa/assignment-1/looking-glass.txt

In [28]:
!wget https://s3.amazonaws.com/2018-dmfa/assignment-1/looking-glass.txt

--2023-09-11 15:57:44--  https://s3.amazonaws.com/2018-dmfa/assignment-1/looking-glass.txt
Resolving s3.amazonaws.com (s3.amazonaws.com)... 54.231.204.16, 16.182.42.56, 52.216.210.224, ...
Connecting to s3.amazonaws.com (s3.amazonaws.com)|54.231.204.16|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 193607 (189K) [text/plain]
Saving to: ‘looking-glass.txt.1’

looking-glass.txt.1 100%[===================>] 189.07K  --.-KB/s    in 0.002s  

2023-09-11 15:57:44 (91.7 MB/s) - ‘looking-glass.txt.1’ saved [193607/193607]



In [31]:
assert 'looking-glass.txt' in os.listdir('.')

Take a look at the next cell.  Will it find the top 25 unique words in *Through the Looking Glass* successfully?

In [27]:
!cat looking-glass.txt | tr -sc '[:alpha:]' '[\n*]' | sort | uniq -c | head -25

      1 
     26 A
      1 ACTUAL
      2 AGED
      2 AGREE
      1 AGREEMENT
      1 AK
      3 AKA
      1 ALICE
      4 ALL
      1 ALONE
      1 ALWAYS
      5 AM
      5 AND
      1 ANSWERS
      4 ANY
      2 ANYTHING
      5 ARE
      1 AS
      2 ASCII
      1 AT
      1 About
      3 Addition
      1 Additional
      7 After
uniq: write error: Broken pipe


Describe what needs to be done to the previous cell to get it to work correctly.  **Describe it using words**, explaining the issues, rather than using shell commands!

**EDIT THIS CELL** WITH YOUR ANSWER HERE
It should sort by the numbers for the top 25 unique words, but now it sorts alphabetically. So we need make upper words to lower and sort the number of top 25 unique words in descening order in order to recieve the correct output


Okay, now implement your solution using shell commands with a pipeline.

In [5]:
!cat looking-glass.txt | tr -sc '[:alpha:]' '[\n*]' |  tr '[:upper:]' '[:lower:]' | sort | uniq -c | sort -rn| head -25

   1775 the
    975 and
    819 a
    817 to
    686 you
    681 it
    660 i
    604 of
    544 she
    473 said
    466 in
    455 alice
    374 that
    358 was
    326 as
    323 t
    309 s
    272 her
    268 with
    260 on
    245 at
    226 he
    225 for
    210 all
    208 but
sort: write failed: 'standard output': Broken pipe
sort: write error
